File: frontend/src/api/todos.ts
import { api } from "./client";

export interface Todo {
  id: number;
  title: string;
  completed: boolean;
  created_at: string;
}

export interface TodoListResponse {
  items: Todo[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListTodosParams {
  page?: number;
  page_size?: number;
  q?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
}

export function listTodos(params?: ListTodosParams): Promise<TodoListResponse> {
  const search = new URLSearchParams();
  if (params?.page) search.set("page", String(params.page));
  if (params?.page_size) search.set("page_size", String(params.page_size));
  if (params?.q) search.set("q", params.q);
  if (params?.sort_by) search.set("sort_by", params.sort_by);
  if (params?.sort_dir) search.set("sort_dir", params.sort_dir);
  const qs = search.toString();
  return api.get<TodoListResponse>(`/api/todos${qs ? `?${qs}` : ""}`);
}

export function createTodo(title: string): Promise<Todo> {
  return api.post<Todo>("/api/todos", { title });
}

export function deleteTodo(id: number): Promise<void> {
  return api.delete<void>(`/api/todos/${id}`);
}

File: backend/app/api/todos.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.engine import get_db
from app.schemas.todo import TodoCreate, TodoListResponse, TodoOut
from app.services.todo import TodoService

router = APIRouter()


@router.get("/todos", response_model=TodoListResponse)
def list_todos(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: str | None = Query(None),
    sort_by: str = Query("id"),
    sort_dir: Literal["asc", "desc"] = Query("asc"),
    db: Session = Depends(get_db),
):
    service = TodoService(db)
    return service.list_todos(
        page=page, page_size=page_size, q=q, sort_by=sort_by, sort_dir=sort_dir
    )


@router.post("/todos", response_model=TodoOut, status_code=201)
def create_todo(data: TodoCreate, db: Session = Depends(get_db)):
    service = TodoService(db)
    return service.create_todo(data)


@router.delete("/todos/{todo_id}", status_code=204)
def delete_todo(todo_id: int, db: Session = Depends(get_db)):
    service = TodoService(db)
    deleted = service.delete_todo(todo_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Todo not found")
